### PR TITLE
Param Refactor ALL

### PR DESCRIPTION
**Pull request by mlzummo**
@GordonSmith @jbrundage @anmoljagetia 

Please Review

Signed-off-by: Mathew Zummo <mzummo@gmail.com>

### CODE REVIEW

File: src/amchart/Area.js [hunk 1]
```diff
@@ -16,14 +16,18 @@
...
     Area.prototype.implements(INDChart.prototype);
-    
-    Area.prototype.publish("paletteID", "Dark2", "set", "Palette ID", Area.prototype._palette.switch());
 
-    Area.prototype.publish("globalTooltipText","[[category]]: [[value]]", "string", "Tooltip Text");
-    Area.prototype.publish("graphTooltipText",["[[category]]: [[value]]"], "array", "Tooltip Text Array");
+    /**
+     * Publish Params Common To Other Libraries
+     */
+    Area.prototype.publish("paletteID", "Dark2", "set", "Palette ID", Area.prototype._palette.switch());
```
Comment (GordonSmith) [inline src/amchart/Area.js]: Wrong default value.

File: src/amchart/Bar.js [hunk 2]
```diff
@@ -16,22 +16,28 @@
     Bar.prototype = Object.create(CommonSerial.prototype);
     Bar.prototype.implements(INDChart.prototype);
     
+    /**
+     * Publish Params Common To Other Libraries
+     */   
     Bar.prototype.publish("paletteID", "Dark2", "set", "Palette ID", Bar.prototype._palette.switch());
```
Comment (GordonSmith) [inline src/amchart/Bar.js]: I know this is not part of this pull request, but its the wrong default palette.

**Comment by GordonSmith** [comment]
All paletteID's should default to "default"

**Comment by GordonSmith** [comment]
What is a tag of "TODO2" mean?  It will should be removed before pulling.

File: src/amchart/Bar.js [hunk 3]
```diff
@@ -16,22 +16,28 @@
...
+    Bar.prototype.publish("Depth3D", 20, "number", "3D Depth (px)",null,{tags:['Basic','TODO2']});
+    Bar.prototype.publish("Angle3D", 45, "number", "3D Angle (Deg)",null,{tags:['Basic','TODO2']});
 
-    Bar.prototype.publish("isStacked", false, "boolean", "Stacked");
-    Bar.prototype.publish("stackType", "regular", "set", "Stack Type",["none","regular","100%"]);
+    Bar.prototype.publish("stackType", "regular", "set", "Stack Type",["none","regular","100%"],{tags:['Basic','TODO2']});
 
-    Bar.prototype.publish("globalTooltipText","[[category]]([[title]]): [[value]]", "string", "Tooltip Text");
-    Bar.prototype.publish("graphTooltipText",["[[category]]([[title]]): [[value]]"], "array", "Tooltip Text");
+    Bar.prototype.publish("tooltipText","[[category]]([[title]]): [[value]]", "string", "Tooltip Text",null,{tags:['Intermediate','TODO2']});
```
Comment (GordonSmith) [inline src/amchart/Bar.js]: Should this be "tooltipPattern" or "tootltipTemplate" rather than tooltiptext?

File: src/amchart/Bar.js [hunk 4]
```diff
@@ -41,14 +47,21 @@
...
+        // Stacked
+        if(this.isStacked()){
+            this._chart.valueAxes[0].stackType = this.stackType();
+        } else {
+            this._chart.valueAxes[0].stackType = "none";
+        }
+        
         // Color Palette
-        if(this.paletteGrouping() === "Main"){
+        if(this.paletteGrouping() === "By Category"){
```
Comment (GordonSmith) [inline src/amchart/Bar.js]: switch(this.paletteGrouping()) would be nicer.

File: src/amchart/Bubble.js [hunk 5]
```diff
@@ -16,9 +16,17 @@
...
+    
+    /**
+     * Publish Params Common To Other Libraries
+     */
     Bubble.prototype.publish("paletteID", "Dark2", "set", "Palette ID", Bubble.prototype._palette.switch());
 
+    /**
+     * Publish Params Unique To This Widget
+     */   
+    Bubble.prototype.publish("tooltipText","[[category]]: [[value]]", "string", "Tooltip Text",null,{tags:['Intermediate','TODO2']});
```
Comment (GordonSmith) [inline src/amchart/Bubble.js]: tooltipTemplate?

File: src/amchart/CommonFunnel.js [hunk 6]
```diff
@@ -19,23 +19,33 @@
...
-    CommonFunnel.prototype.publish("chartScrollbar", false, "boolean", "Chart Scrollbar");
+    CommonFunnel.prototype.publish("showScrollbar", false, "boolean", "Show Chart Scrollbar",null,{tags:['Intermediate','TODO2']});
     
-    CommonFunnel.prototype.publish("startDuration", 0.3, "number", "Start Duration (sec)");
+    CommonFunnel.prototype.publish("startDuration", 0.3, "number", "Start Duration (sec)",null,{tags:['Private','TODO2']});
     
-    CommonFunnel.prototype.publish("Depth3D", 0, "number", "3D Depth (px)");
-    CommonFunnel.prototype.publish("Angle3D", 0, "number", "3D Angle (Deg)");
+    CommonFunnel.prototype.publish("Depth3D", 0, "number", "3D Depth (px)",null,{tags:['Basic','TODO2']});
+    CommonFunnel.prototype.publish("Angle3D", 0, "number", "3D Angle (Deg)",null,{tags:['Basic','TODO2']});
```
Comment (GordonSmith) [inline src/amchart/CommonFunnel.js]: Should all Angle3d and Depth 3D default to the same value(s)?

File: src/amchart/CommonXY.js [hunk 7]
```diff
@@ -57,6 +97,10 @@
         this._chart.theme = "none";
         this._chart.type = "xy";
         this._chart.pathToImages = "//cdn.rawgit.com/cdnjs/cdnjs/master/ajax/libs/amcharts/3.13.0/images/";
```
Comment (GordonSmith) [inline src/amchart/CommonXY.js]: This needs to be fixed / removed...  (amCharts should be able to work on systems with no internet connectivity).

File: src/amchart/FloatingColumn.js [hunk 8]
```diff
@@ -48,7 +55,7 @@
     
     FloatingColumn.prototype.columns = function(colArr) {
         if (!arguments.length) return this._columns;
-        var retVal = CommonSerial.prototype.columns.apply(this, arguments);
+        //var retVal = CommonSerial.prototype.columns.apply(this, arguments);
```
Comment (GordonSmith) [inline src/amchart/FloatingColumn.js]: remove line.

File: src/amchart/Gauge.js [hunk 9]
```diff
@@ -19,33 +19,40 @@
...
+    Gauge.prototype.publish("tickAlpha", 0.2, "number", "Tick Alpha",null,{tags:['Intermediate','TODO2']});
+    Gauge.prototype.publish("valueInterval", 20, "number", "Value Interval",null,{tags:['Advanced','TODO2']});
+    Gauge.prototype.publish("bottomText", "", "string", "Text Along Bottom",null,{tags:['Intermediate','TODO2']});
+    Gauge.prototype.publish("bottomTextYOffset", -20, "number", "Bottom Text Vertical Offset",null,{tags:['Intermediate','TODO2']});
     
-    Gauge.prototype.publish("animateDuration", 2, "number", "");
+    Gauge.prototype.publish("animatationDuration", 2, "number", "Animation Duration (sec)",null,{tags:['Intermediate','TODO2']});
 
-    Gauge.prototype.publish("globalTooltipText","[[category]]([[title]]): [[value]]", "string", "Tooltip Text");
+    //Gauge.prototype.publish("tooltipText","[[category]]([[title]]): [[value]]", "string", "Tooltip Text",null,{tags:['Intermediate','TODO2']});
```
Comment (GordonSmith) [inline src/amchart/Gauge.js]: remove commented code.

File: src/amchart/Pie.js [hunk 10]
```diff
@@ -57,7 +59,7 @@
         this._chart.radius = this.radius();
         this._chart.pathToImages = "//cdn.rawgit.com/cdnjs/cdnjs/master/ajax/libs/amcharts/3.13.0/images/";
```
Comment (GordonSmith) [inline src/amchart/Pie.js]: needs to be removed/fixed

File: src/c3chart/Area.js [hunk 11]
```diff
@@ -14,16 +14,24 @@
...
+    /**
+     * Publish Params Unique To This Library
+     */
+    
+    Area.prototype.enter = function (domNode, element) {       
         CommonND.prototype.enter.apply(this,arguments);
     }
     
-    Area.prototype.update = function (domNode, element) {
+    Area.prototype.update = function (domNode, element) {
```
Comment (GordonSmith) [inline src/c3chart/Area.js]: remove trailing spaces

File: src/c3chart/Area.js [hunk 12]
```diff
@@ -35,7 +43,11 @@
         element.selectAll(".c3-line").style({ 
             "stroke-width": this.lineWidth()+"px", 
             "stroke-opacity": this.lineOpacity(), 
-            "stroke-dasharray": this.dashedLine().toString(), 
+            "stroke-dasharray": this.lineDashStyle().toString(),
```
Comment (GordonSmith) [inline src/c3chart/Area.js]: remove trailing comma

File: src/c3chart/CommonND.js [hunk 13]
```diff
@@ -28,28 +28,42 @@
...
+    CommonND.prototype.publish("axisLineWidth", 1, "number", "Axis Line Width",null,{tags:['Intermediate','TODO2']});
+    
+    CommonND.prototype.publish("xAxisBaselineColor", "#000", "html-color", "X Axis Baseline Color",null,{tags:['Basic','TODO2']});
+    CommonND.prototype.publish("yAxisBaselineColor", "#000", "html-color", "Y Axis Baseline Color",null,{tags:['Basic','TODO2']});
+
+    CommonND.prototype.publish("xAxisFontColor", "#000", "html-color", "X Axis Text Font Color",null,{tags:['Basic','TODO2']});
+    CommonND.prototype.publish("yAxisFontColor", "#000", "html-color", "Y Axis Text Font Color",null,{tags:['Basic','TODO2']});
+
+    CommonND.prototype.publish("axisFontSize", 10, "number", "X/Y Axis Text Font Size",null,{tags:['Basic','TODO2']});
+    CommonND.prototype.publish("axisFontFamily", "sans-serif", "string", "X/Y Axis Text Font Name",null,{tags:['Basic','TODO2']});
```
Comment (GordonSmith) [inline src/c3chart/CommonND.js]: should default to null?

File: src/c3chart/c3.css [hunk 14]
```diff
@@ -0,0 +1,162 @@
+/*-- Chart --*/
```
Comment (GordonSmith) [inline src/c3chart/c3.css]: This file should not be included.

**Comment by GordonSmith** [comment]
General rule of thumb - any property which can be set via a stylesheet, should be set in to default null in our published properties?

File: src/google/Area.js [hunk 15]
```diff
@@ -15,79 +15,95 @@
...
+    Area.prototype.publish("fillOpacity", null, "number", "Opacity of The Fill Color",null,{tags:['Intermediate']});
+    
+    Area.prototype.publish("axisFontSize", null, "number", "X/Y Axis Label Font Size",null,{tags:['Basic']});
+    Area.prototype.publish("axisFontFamily", null, "string", "X/Y Axis Label Font Name",null,{tags:['Basic']});
+    
+    Area.prototype.publish("xAxisFontColor", null, "html-color", "X Axis Text Font Color",null,{tags:['Basic']});
+    Area.prototype.publish("yAxisFontColor", null, "html-color", "X Axis Text Font Color",null,{tags:['Basic']});
+    
+    Area.prototype.publish("xAxisBaselineColor", "#000000", "html-color", "Specifies The Color of The Baseline For The Horizontal Axis",null,{tags:['Intermediate']});
+    Area.prototype.publish("yAxisBaselineColor", "#000000", "html-color", "Specifies The Color of The Baseline For The Vertical Axis",null,{tags:['Intermediate']});
```
Comment (GordonSmith) [inline src/google/Area.js]: Should be null

File: src/google/Area.js [hunk 16]
```diff
@@ -15,79 +15,95 @@
...
     
-    Area.prototype.publish("hAxisTicks", [], "array", "Replaces the automatically generated X-axis ticks with the specified array");
+    Area.prototype.publish("xAxisFormat", "", "string", "Format String For Numeric Axis Labels", ["","decimal","scientific","currency","percent","short","long"],{tags:['Intermediate']});
+    Area.prototype.publish("yAxisFormat", "", "string", "Format String For Numeric Axis Labels", ["","decimal","scientific","currency","percent","short","long"],{tags:['Intermediate']});
+    
+    Area.prototype.publish("xAxisGridlinesCount", 5, "number", "The Number of Horizontal Gridlines Between Two Regular Gridlines",null,{tags:['Intermediate']});
+    Area.prototype.publish("yAxisGridlinesCount", 5, "number", "The Number of Vertical Gridlines Between Two Regular Gridlines",null,{tags:['Intermediate']});
+    
+    Area.prototype.publish("xAxisGridlinesColor", "#CCC", "html-color", "The Color of The Horizontal Gridlines Inside The Chart Area",null,{tags:['Basic']});
+    Area.prototype.publish("yAxisGridlinesColor", "#CCC", "html-color", "The Color of The Vertical Gridlines Inside The Chart Area",null,{tags:['Basic']});
```
Comment (GordonSmith) [inline src/google/Area.js]: Should be null

File: src/google/Area.js [hunk 17]
```diff
@@ -15,79 +15,95 @@
...
-    Area.prototype.publish("vAxisTitleTextStyleFontSize", null, "number", "Vertical axis titletext style (Font Size)");
-    
-    Area.prototype.publish("vAxisMaxValue", null, "number", "Moves the max value of the vertical axis to the specified value");
-    Area.prototype.publish("vAxisMinValue", null, "number", "Moves the min value of the vertical axis to the specified value");
+    Area.prototype.publish("xAxisMinorGridlinesCount", 0, "number", "The Number of Horizontal Minor Gridlines Between Two Regular Gridlines",null,{tags:['Intermediate']});
+    Area.prototype.publish("yAxisMinorGridlinesCount", 0, "number", "The Number of Vertical Minor Gridlines Between Two Regular Gridlines",null,{tags:['Intermediate']});
+    
+    Area.prototype.publish("xAxisMinorGridlinesColor", "#FFFFFF", "html-color", "The Color of The Horizontal Minor Gridlines Inside The Chart Area",null,{tags:['Intermediate']});
+    Area.prototype.publish("yAxisMinorGridlinesColor", "#FFFFFF", "html-color", "The Color of The Vertical Minor Gridlines Inside The Chart Area",null,{tags:['Intermediate']});
```
Comment (GordonSmith) [inline src/google/Area.js]: Should be null.

File: src/layout/Surface.js [hunk 18]
```diff
@@ -23,29 +39,32 @@
...
 
     Surface.prototype.update = function (domNode, element) {
         HTMLWidget.prototype.update.apply(this, arguments);
+        
+        element
+            .style("border-width",this.borderWidth())
+            .style("border-color",this.borderColor())
+            .style("border-radius",this.borderRadius())
+            .style("background-color",this.backgroundColor())
+
```
Comment (GordonSmith) [inline src/layout/Surface.js]: Missing ";"

File: src/amchart/CommonFunnel.js (hunk 6 above)
Comment (mlzummo) [inline src/amchart/CommonFunnel.js]: not sure what you mean exactly.

File: src/amchart/CommonXY.js (hunk 7 above)
Comment (mlzummo) [inline src/amchart/CommonXY.js]: itll work but its going to throw errors saying it cant find the images ... going to remove

File: src/other/Persist.js [hunk 19]
```diff
@@ -88,7 +88,8 @@
 
         deserialize: function (state, callback) {
             var context = this;
-            var path = "src/" + state.__class.split("_").join("/");
+            var classSplit = state.__class.split(' ');
+            var path = "src/" + classSplit[classSplit.length-1].split("_").join("/");
```
Comment (mlzummo) [inline src/other/Persist.js]: this needs to be fixed/reverted

File: src/other/PropertyEditor.js [hunk 20]
```diff
@@ -241,11 +242,12 @@
             });
         }
         //Update tables based on "group by" setting
-        if (this.param_grouping() === "By Param") {
+        if (this.paramGrouping() === "By Param") {
+
```
Comment (mlzummo) [inline src/other/PropertyEditor.js]: remove extra space

File: src/other/PropertyEditor.js [hunk 21]
```diff
@@ -336,10 +338,11 @@
                 var tbody = d3.select(this);
                 var rows;
                 //  Columns  ---
+
```
Comment (mlzummo) [inline src/other/PropertyEditor.js]: remove extra space?

File: src/amchart/CommonSerial.js [hunk 22]
```diff
@@ -10,145 +10,195 @@
...
-                gObj.lineColorField = "linecolor";
-            }
-        }
+
+//        if (typeof(context.paletteGrouping) === 'function') {
+//            if(context.paletteGrouping() === "By Category"){
+//                gObj.colorField = "color";
+//                gObj.lineColorField = "linecolor";
+//            }
+//        }
```
Comment (mlzummo) [inline src/amchart/CommonSerial.js]: remove this commented section

**Comment by mlzummo** [comment]
need to fix bug in showLabel/showValueLabel/showLegend

**Comment by mlzummo** [comment]
@GordonSmith Ready to be pulled

**Comment by jbrundage** [comment]
Looks good to me

**Comment by mlzummo** [comment]
do not pull yet

**Comment by mlzummo** [comment]
ready to pull ... left the second commit because it adds test/test-no-amd and several other fixes

**Comment by GordonSmith** [comment]
Closing in favour of:  https://github.com/hpcc-systems/Visualization/pull/394

**Pull request closed without merging**